mediatek: add support for Buffalo WSR-6000AX8 and AX8P

ssyysy2021 opened the pull request:
This commit adds support for Buffalo WSR-6000AX8 and AX8P models.

Hardware
--------
SOC:       MediaTek MT7986B
RAM:       512MB
FLASH:     128MB SPI-NAND (Winbond)
WIFI 2.4G: (Embedded in SOC) b/g/n/ax, MIMO 4x4
WIFI 5G:   (Embedded in SOC) a/n/ac/ax, MIMO 4x4
ETHERNET:  2.5GbE MaxLinear GPY211C (eth1: WAN)
ETHERNET:  MediaTek MT7531AE 3xGbE (eth0: LAN1, LAN2, LAN3)
UART:      3.3V 115200 8N1

Serial(UART) Pin Layout
-----------------------

&nbsp;&nbsp;&nbsp;+---+----+-----+------+
J4 |  RX   |  TX   |  GND  |(3.3V) |
&nbsp;&nbsp;&nbsp;+---+----+-----+------+
           (Bottom Side)

MAC Address Table
-----------------
eth0/LAN:  board_data 0x4(text)
eth1/WAN:  board_data 0x4(text)
WIFI 2.4G: eth0 + 2
WIFI 5G:   eth0 + 9

Installation
------------

1. Set up a TFTP server with the IP address "192.168.11.2".
2. Rename the OpenWrt initramfs image to "linux.ubi-recovery" and place it in the TFTP server's root directory.
3. While holding down the AOSS button, power on the WSR-6000AX8 (or AX8P).
4. The device will automatically download the initramfs image from the TFTP server and boot into it.
5. Once booted, run "sysupgrade -n" using the OpenWrt sysupgrade image to complete the installation.

Recovery
--------

1. Decrypt the official firmware image using "buffalo-enc".
   example:   $ buffalo-enc -d -i wsr_6000ax8_jp_100 -o wsr_6000ax8_jp_100.dec \ -l -O 0xc8
2. Run "sysupgrade -F -n" with the decrypted image.

ssyysy2021 commented:
@musashino205 Thank you for informing me about 'bootargs-append' on [Twitter](https://twitter.com/musashino_205/status/1679494497749839875).  
  
 I will be closing this pull request as I have other aspects that I would like to enhance.

musashino205 commented:
If you make additional changes and continue this contribution for that device, please use ```git push -f``` to update this PR instead of opening a new PR.
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#do_not_multi-post_or_re-post_prs

BTW, please use the target name instead of subtarget name for the commit prefix. (```mediatek: ```)
https://openwrt.org/submitting-patches#submission_guidelines

ssyysy2021 commented:
I have reopened the pull request. Please review my code.

dragouf commented:
Is there a chance this will be merged one day? I would be very interested. Anyway, thanks for your work

hauke commented:
Please rebase this on top of current main branch and test it again.

musashino205 commented on package/boot/uboot-envtools/files/mediatek_filogic:
```diff
@@ -101,6 +101,7 @@ glinet,gl-mt6000)
 glinet,gl-mt3000)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x80000" "0x20000"
 	;;
+buffalo,wsr-6000ax8|\
```
Please sort alphabetically. (this block should be placed before `cmcc,rax3000m`)

ssyysy2021 replied:
Thank you for pointing that out. I've fixed it.

musashino205 commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dts:
```diff
@@ -0,0 +1,365 @@
+		mediatek,nmbm;
+		mediatek,bmt-max-ratio = <1>;
+		mediatek,bmt-max-reserved-blocks = <64>;
+		mediatek,bmt-remap-range = <0x0 0x580000>,
+								   <0x3980000 0x7480000>;
```
wrong indentation?

ssyysy2021 replied:
Done

hauke commented:
Please rebase this on top of current main branch and test it again.

ssyysy2021 commented:
> Please rebase this on top of current main branch and test it again.
I've rebased the changes onto the latest main branch and re-tested — apologies for the delay.
Additionally, I've added support for another model with nearly identical hardware.

ssyysy2021 commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dts:
```diff
@@ -0,0 +1,365 @@
+		compatible = "spi-nand";
+		reg = <0>;
+
+		spi-max-frequency = <20000000>;
```
Thanks for the review.  I've made the changes.

hauke commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dtsi:
```diff
@@ -0,0 +1,382 @@
+	};
+
+	memory {
+		reg = <0 0x40000000 0 0x20000000>;
```
```suggestion
		reg = <0 0x40000000 0 0x20000000>;
		device_type = "memory";
```

hauke commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dtsi:
```diff
@@ -0,0 +1,382 @@
+		bootargs-override = "ubi.mtd=ubi";
+	};
+
+	memory {
```
```suggestion
	memory@40000000 {
```

ssyysy2021 replied:
Fixed.

hauke commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dtsi:
```diff
@@ -0,0 +1,382 @@
+			       "WF1_HB2", "WF1_HB3", "WF1_HB4", "WF1_HB0",
+			       "WF1_HB5", "WF1_HB6", "WF1_HB7", "WF1_HB8",
+			       "WF1_TOP_CLK", "WF1_TOP_DATA";
+			drive-strength = <4>;
```
```suggestion
			drive-strength = <MTK_DRIVE_4mA>;
```

and include this in the top of the file:
```
#include <dt-bindings/pinctrl/mt65xx.h>
```

ssyysy2021 replied:
Fixed.

hauke commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dtsi:
```diff
@@ -0,0 +1,382 @@
+			factory: partition@180000 {
+				label = "Factory";
+				reg = <0x180000 0x200000>;
+				read-only;
```
Add an nvram for the eeprom:

```suggestion
				read-only;

				nvmem-layout {
					compatible = "fixed-layout";
					#address-cells = <1>;
					#size-cells = <1>;

					eeprom_factory_0: eeprom@0 {
						reg = <0x0 0x1000>;
					};
				};
```

hauke commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dtsi:
```diff
@@ -0,0 +1,382 @@
+		};
+		conf-pd {
+			pins = "SPI2_CLK", "SPI2_MOSI", "SPI2_MISO";
+			drive-strength = <8>;
+			mediatek,pull-down-adv = <0>; /* bias-disable */
```
Please adapt the dts like it was done in https://github.com/openwrt/openwrt/pull/19741

ssyysy2021 replied:
Fixed.

hauke commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dtsi:
```diff
@@ -0,0 +1,382 @@
+	pinctrl-0 = <&wf_2g_5g_pins>;
+	pinctrl-1 = <&wf_dbdc_pins>;
+
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
Please check that the mac address assignment still works like expected after these changes.

ssyysy2021 replied:
I've fixed both issues and confirmed that the MAC address is correctly reflected.

ssyysy2021 commented on target/linux/mediatek/dts/mt7986b-buffalo-wsr-6000ax8.dtsi:
```diff
@@ -0,0 +1,382 @@
+			       "WF1_HB2", "WF1_HB3", "WF1_HB4", "WF1_HB0",
+			       "WF1_HB5", "WF1_HB6", "WF1_HB7", "WF1_HB8",
+			       "WF1_TOP_CLK", "WF1_TOP_DATA";
+			drive-strength = <4>;
```
Fixed.

hauke commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -583,6 +583,28 @@ endif
 endef
 TARGET_DEVICES += bananapi_bpi-r4-poe
 
+define Device/buffalo_wsr-6000ax8-common
+  DEVICE_VENDOR := Buffalo
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7986-firmware mt7986-wo-firmware
+  IMAGE_SIZE := 26624k
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+endef
+
+define Device/buffalo_wsr-6000ax8
+  DEVICE_MODEL := WSR-6000AX8
+  DEVICE_DTS := mt7986b-buffalo-wsr-6000ax8
+  $(call Device/buffalo_wsr-6000ax8-common)
+endef
+TARGET_DEVICES += buffalo_wsr-6000ax8
+
+define Device/buffalo_wsr-6000ax8p
+  DEVICE_MODEL := WSR-6000AX8P
+  DEVICE_DTS := mt7986b-buffalo-wsr-6000ax8p
+  $(call Device/buffalo_wsr-6000ax8-common)
+endef
+TARGET_DEVICES += buffalo_wsr-6000ax8p
+
```
Do you need two different images?

I would prefer only one image:
```suggestion
define Device/buffalo_wsr-6000ax8
  DEVICE_MODEL := WSR-6000AX8
  DEVICE_VENDOR := Buffalo
  DEVICE_ALT0_MODEL := WSR-6000AX8P
  DEVICE_ALT0_VENDOR := Buffalo
  DEVICE_DTS_DIR := ../dts
  DEVICE_DTS := mt7986b-buffalo-wsr-6000ax8
  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7986-firmware mt7986-wo-firmware
  IMAGE_SIZE := 26624k
  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
endef
TARGET_DEVICES += buffalo_wsr-6000ax8

```

castiel652 commented:
Can you share the output of `hexdump -Cs 0x19a -n 1 /dev/mtd2`

ssyysy2021 commented:
> Can you share the output of `hexdump -Cs 0x19a -n 1 /dev/mtd2`
@castiel652 It's 0x07.

ssyysy2021 committed:
mediatek: add support for Buffalo WSR-6000AX8 and AX8P

This commit adds support for Buffalo WSR-6000AX8 and AX8P models.

Hardware
--------
SOC:       MediaTek MT7986B
RAM:       512MB
FLASH:     128MB SPI-NAND (Winbond)
WIFI 2.4G: (Embedded in SOC) b/g/n/ax, MIMO 4x4
WIFI 5G:   (Embedded in SOC) a/n/ac/ax, MIMO 4x4
ETHERNET:  2.5GbE MaxLinear GPY211C (eth1: WAN)
ETHERNET:  MediaTek MT7531AE 3xGbE (eth0: LAN1, LAN2, LAN3)
UART:      3.3V 115200 8N1

Serial(UART) Pin Layout
-----------------------

   +-------+-------+-------+-------+
J4 |  RX   |  TX   |  GND  |(3.3V) |
   +-------+-------+-------+-------+
           (Bottom Side)

MAC Address Table
-----------------
lan1-3:    board_data 0x4(text)
eth1/WAN:  board_data 0x4(text)
WIFI 2.4G: lan1 + 2
WIFI 5G:   lan1 + 9

Installation
------------

1. Set up a TFTP server with the IP address "192.168.11.2".

2. Rename the OpenWrt initramfs image to "linux.ubi-recovery" and place it
   in the TFTP server's root directory.
3. While holding down the AOSS button, power on the WSR-6000AX8 (or AX8P).

4. The device will automatically download the initramfs image
   from the TFTP server and boot into it.
5. Once booted, run "sysupgrade -n" using the OpenWrt sysupgrade image
   to complete the installation.

Recovery
--------

1. Decrypt the official firmware image using "buffalo-enc".

   example:
   $ buffalo-enc -d -i wsr_6000ax8_jp_100 -o wsr_6000ax8_jp_100.dec \
     -l -O 0xc8

2. Run "sysupgrade -F -n" with the decrypted image.

Signed-off-by: Shin Sato <shin.sugar.ssyysy2021@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/13107
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
